server/pulp/server/controllers/repository.py

```diff
@@ -105,6 +105,46 @@ def find_repo_content_units(
             yield_count += 1
 
 
+def find_units_not_downloaded(repository):
+    """
+    Find content units that have not been fully downloaded.
+
+    :param repository: A repository.
+    :type repository: model.Repository
+    :return: The requested units.
+    :rtype: generator
+    """
+    types = {}
+    for association in model.RepositoryContentUnit.objects(repo_id=repository.repo_id):
+        unit_ids = types.setdefault(association.unit_type_id, [])
+        unit_ids.append(association.unit_id)
+    for type_id, unit_ids in types.items():
+        _model = plugin_api.get_unit_model_by_id(type_id)
+        for unit in _model.objects(id__in=unit_ids, downloaded=False):
```

Comment:
check out ``pulp.plugins.util.misc.paginate``, which would avoid creating a gigantic query here that might exceed size limits.